### DISCUSSION TITLE
shorter collatz sequence?

### DISCUSSION THREAD

**Discussion by tungster24** [abrudz/aplcart#120] [Ideas]
I came up with a shorter way to generate the collatz sequence. I'm not sure if it can be as easily used, but if anyone is willing to take a look here it is:
`l←{⊃⍵=1:⍵⋄⍵,∇{2|⍵:1+3×⍵⋄⍵÷2}⊃⍵}`

**Comment by abrudz** [comment]
Thank you for this. Inspired by it, and [with some help from a code colfer](https://chat.stackexchange.com/transcript/message/65647019#65647019), I managed to come up with `{∇⍣(1≠n)⊢⍵,n←2(|+÷⍨×1+5×|)⊢/⍵}` which I like because it has no branching at all, but it is slight faster to use a branch than `⍣` (here for `n←670617279`):
```
{∇⍣(1≠n)⊢⍵,n←2(|+÷⍨×1+5×|)⊢/⍵} n → 3.6E¯3 |  0% ⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕ 
{1=l←⊢/⍵:⍵⋄∇⍵,2(|+÷⍨×1+5×|)l} n  → 3.4E¯3 | -7% ⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕    
```
Anyway, there are basically two parts; recursion vs iteration and direct formula vs branching. We can compare the performance of all four as follows:
```
{1=l←⊢/⍵:⍵⋄∇⍵,2(|+÷⍨×1+5×|)l} n     → 3.4E¯3 |   0% ⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕ 
{1=l←⊢/⍵:⍵⋄∇⍵,{2|⍵:1+3×⍵⋄⍵÷2}l} n   → 1.8E¯3 | -46% ⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕                   
{r⊣2(|+÷⍨×1+5×|)⍣{1=r,←⍵}⍵⊣r←⍬} n   → 2.0E¯3 | -40% ⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕                 
{r⊣{2|⍵:1+3×⍵⋄⍵÷2}⍣{1=r,←⍵}⍵⊣r←⍬} n → 1.5E¯3 | -55% ⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕                       
               
```
So, the current solution (iterative with branching) is fastest.

---
Note that these recursive solutions take advantage of TCO, while yours doesn't, as it needs to wait for `∇` to return before it can compute `⍵,∇` and we can see the impact of that on performance, even if revese the catenation:
```
{1=f←⊃⍵:⍵⋄⍵,∇{2|⍵:1+3×⍵⋄⍵÷2}f} n   → 7.6E¯3 |   0% ⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕ 
{1=l←⊢/⍵:⍵⋄∇⍵,{2|⍵:1+3×⍵⋄⍵÷2}l} n  → 1.9E¯3 | -76% ⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕                               
⌽{1=f←⊃⍵:⍵⋄∇⍵,⍨{2|⍵:1+3×⍵⋄⍵÷2}f} n → 2.0E¯3 | -74% ⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕                               
```
---
And if we're after golfing, an adaption of [this](https://codegolf.stackexchange.com/questions/12177/collatz-conjecture-oeis-a006577#comment374535_12182) is the shortest I've seen:
```
{1=f←⊃⍵:⍵⋄⍵,∇⊃f⌽0 1+.5 3×f}
```
